### ISSUE TITLE
Reducing memory usage for large objects

### ISSUE THREAD

**Issue by Tornhoof** [Tornhoof/SpanJson#42]
**Introduction**
SpanJson reads/writes the data only as a block and if the data to be read/written is large enough to allocate buffers on the LOH the performance and memory usage will suffer.
As long as SpanJson is not working with large arrays on the LOH the GC is happy and we're happy.

**Solution**
For the streaming async API (Stream/Textwriter/Textreader) a way needs to be found to flush the data at specific points to make sure we don't allocate on the LOH and instead reuse a smaller buffer again.

***Writing***
* We have specific points where we can flush, specifically e.g. after an array, list, object etc.
* It's not necessary to flush after most normal data types, only string and related writer methods
* We already track the size of the fully serialized/deserialized object, we can use that to decide if we want to run the flush approach or the normal block approach.

**POC for writing** 

I've implemented that idea for writing a large string list, with N "Hello World and Universe" strings.

***Questions***
- How often do we flush?
Obviously before we reach the end of the buffer, this needs some calculations for string et. al.

I also guess it's a function based on buffer size and underlying stream performance,
my current assumption is flush shortly before we reach the LOH size (e.g.~85kb).

``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Xeon CPU E5-1620 0 3.60GHz, 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.301
  [Host]     : .NET Core 2.1.1 (CoreCLR 4.6.26606.02, CoreFX 4.6.26606.05), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.1 (CoreCLR 4.6.26606.02, CoreFX 4.6.26606.05), 64bit RyuJIT


```
|      Method | Count |        Mean |      Error |     StdDev |      Median |    Gen 0 |    Gen 1 |    Gen 2 |  Allocated |
|------------ |------ |------------:|-----------:|-----------:|------------:|---------:|---------:|---------:|-----------:|
| **NewWayUtf16** |  **1000** |    **82.70 us** |  **2.1447 us** |  **3.1436 us** |    **81.06 us** |   **1.7090** |        **-** |        **-** |    **9.23 KB** |
|  NewWayUtf8 |  1000 |   101.63 us |  1.2475 us |  1.1669 us |   101.20 us |   0.7324 |        - |        - |    4.02 KB |
| OldWayUtf16 |  1000 |    77.22 us |  0.2569 us |  0.2403 us |    77.13 us |  12.4512 |        - |        - |   64.02 KB |
|  OldWayUtf8 |  1000 |   106.23 us |  0.4006 us |  0.3345 us |   106.14 us |   6.2256 |        - |        - |   32.02 KB |
| **NewWayUtf16** | **10000** |   **796.14 us** |  **2.0300 us** |  **1.8989 us** |   **795.83 us** |   **0.9766** |        **-** |        **-** |    **9.23 KB** |
|  NewWayUtf8 | 10000 | 1,010.74 us | 11.9164 us | 11.1466 us | 1,005.98 us |        - |        - |        - |    4.02 KB |
| OldWayUtf16 | 10000 |   924.31 us |  6.7135 us |  5.2415 us |   923.75 us | 332.0313 | 332.0313 | 332.0313 | 1024.02 KB |
|  OldWayUtf8 | 10000 | 1,124.74 us |  1.5156 us |  1.2656 us | 1,124.38 us | 166.0156 | 166.0156 | 166.0156 |  512.02 KB |

**Comment by Tornhoof** [comment]
I think for writing it works, for Reading it's basically the same problem just string et al. is quite a bit more work,  because the we might need to resize the buffer to be able to read the whole string into it.

**Open Questions**
***when to flush/reframe***
Reframe is basically moving the sliding window further on stream)
- We need to flush/reframe before writing/reading strings et. al. to make sure the buffer is properly sized
- Every other datatype is at most a fixed size, so we shouldn't even attempt to try to flush/reframe after each one.
- For lists, enumerables etc. we might want to try to flush/reframe after buffersize/fixed size entries.
- For complex types we should flush/reframe after each type, just to be sure and we might need to flush/reframe if it has a shitload of value type properties (those with fixed size),e.g. after buffersize/fixed size properties. (we should rarely hit that).

Buffer size is another question, assuming we only run the code path for if the estimated object size is big enough, then obviously we can use a fairly large buffer, 2048 or 4096 should be ok.

**Comment by Tornhoof** [comment]
Long story short:
* Having a second set of Jsonwriter/Reader works, but is very invasive, meaning all the complex formatters need extra logic for them.
* Pushing the streaming Logic into the Json Writer/Reader primarily reduces the Performance of the normal Block based approach way too much to be acceptable.
* Having a custom indexer maintaining the Slide/Flush Logic is several times slower.

TL;DR: The first approach is only useful, If it supports async read/write, otherwise the api Surface increases too much.
The other approaches all reduce the Performance of the Block based Version way too much.
